dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorBase.java

```diff
@@ -57,33 +48,35 @@ public List<Event> visitInitLock(InitLock e) {
 
     @Override
     public List<Event> visitLock(Lock e) {
-        IntegerType type = (IntegerType) e.getAccessType(); // TODO: Boolean should be sufficient
+        Type type = types.getBooleanType();
         Register dummy = e.getFunction().newRegister(type);
-        Expression zero = expressions.makeZero(type);
-        Expression one = expressions.makeOne(type);
+        Expression one = expressions.makeOne((IntegerType) e.getAccessType());
         String mo = e.getMo();
 
         Load rmwLoad = newRMWLoadWithMo(dummy, e.getAddress(), mo);
+        Label spinLoopHead = newLabel("__spinloop_head");
         return eventSequence(
+                spinLoopHead,
                 rmwLoad,
-                newTerminator(expressions.makeNEQ(dummy, zero)),
+                newJump(dummy, spinLoopHead),
                 newRMWStoreWithMo(rmwLoad, e.getAddress(), one, mo)
         );
     }
 
     @Override
     public List<Event> visitUnlock(Unlock e) {
-        IntegerType type = (IntegerType) e.getAccessType(); // TODO: Boolean should be sufficient
+        Type type = types.getBooleanType();
         Register dummy = e.getFunction().newRegister(type);
-        Expression zero = expressions.makeZero(type);
-        Expression one = expressions.makeOne(type);
+        Expression zero = expressions.makeZero((IntegerType) e.getAccessType());
         Expression address = e.getAddress();
         String mo = e.getMo();
 
         Load rmwLoad = newRMWLoadWithMo(dummy, address, mo);
+        Label spinLoopHead = newLabel("__spinloop_head");
         return eventSequence(
+                spinLoopHead,
                 rmwLoad,
-                newTerminator(expressions.makeNEQ(dummy, one)),
+                newJump(expressions.makeNot(dummy), spinLoopHead),
```

Comment:
Why does unlock even spin? Unlocking an already unlocked mutex is UB so it would make more sense to put an assertion to check that. Alternatively, if we do not want to check for UB, the load operation can go completely